src/models/document/base-document-content.ts

```diff
@@ -1002,209 +1008,115 @@ export const BaseDocumentContentModel = types
+    // Returns { oldTitle, newTitle }, which are undefined if the title wasn't changed
+    updateDefaultTileTitle(tileId: string) {
+      const tile = self.getTile(tileId);
+      if (tile) {
+        const tileContentInfo = getTileContentInfo(tile.content.type);
+        if (tileContentInfo) {
+          const oldTitle = tile.title;
+          const match = titleMatchesDefault(oldTitle, tileContentInfo.titleBase);
+          if (match) {
+            const newTitle = self.getNewTileTitle(tile.content.type);
+            tile.setTitle(newTitle);
+            return { oldTitle, newTitle };
           }
         }
-        catch(e) {
-          // ignore shared models with errors
-        }
-      });
-    },
-    duplicateTiles(tiles: IDragTileItem[]) {
-      // Determine the row to add the duplicated tiles into
-      const rowIndex = self.getRowAfterTiles(tiles);
-
-      // Find shared models used by tiles being duplicated
-      const sharedModelEntries = Object.values(self.getSharedModelsUsedByTiles(tiles.map(tile => tile.tileId)));
-
-      // Update shared models with new ids
-      interface UpdatedSharedDataSetIds {
-        attributeIdMap: Record<string, string>;
-        caseIdMap: Record<string, string>;
-        dataSetId: string;
-        sharedModelId: string;
       }
+      return { oldTitle: undefined, newTitle: undefined };
+    }
+  }))
+  .actions(self => ({
+    // Copies tiles and shared models into the specified row, giving them all new ids
+    copyTiles(
+      tiles: IDragTileItem[],
+      sharedModelEntries: PartialSharedModelEntry[],
+      rowInfo: IDropRowInfo,
+      insertTileFunction: (updatedTiles: IDropTileItem[], rowInfo: IDropRowInfo) => NewRowTileArray
+    ) {
+      // Update shared models with new ids
       const updatedSharedModelMap: Record<string, UpdatedSharedDataSetIds> = {};
-      const newSharedModelEntries: any[] = [];
+      const newSharedModelEntries: PartialSharedModelEntry[] = [];
       sharedModelEntries.forEach(sharedModelEntry => {
         // For now, only duplicate shared data sets
         if (sharedModelEntry.sharedModel.type === "SharedDataSet") {
           // Determine new ids
-          const sharedDataSet = getSnapshot(sharedModelEntry.sharedModel) as SharedDataSetSnapshotType;
-          const updatedEntry: UpdatedSharedDataSetIds = {
-            attributeIdMap: {},
-            caseIdMap: {},
-            dataSetId: uniqueId(),
-            sharedModelId: uniqueId()
-          };
-          sharedDataSet.dataSet.attributes?.forEach(attr => {
-            updatedEntry.attributeIdMap[attr.id] = uniqueId();
-          });
-          sharedDataSet.dataSet.cases?.forEach(c => {
-            if (c.__id__) updatedEntry.caseIdMap[c.__id__] = newCaseId();
-          });
-          if (sharedDataSet.id) updatedSharedModelMap[sharedDataSet.id] = updatedEntry;
+          const sharedDataSet = sharedModelEntry.sharedModel as SharedDataSetSnapshotType;
```

Comment:
FYI, you should be able to write a typing function like this (which could live in the `SharedDataSet` module):
```typescript
export function isSharedDataSetSnapshot(snap: any): snap is SharedDataSetSnapshotType {
  return snap.type === "SharedDataSet";
}
```
and then call it in line 1044 above and then you wouldn't need to cast here.

Reply:
Looks good -- the cast in 1042 should no longer be necessary.